darwin: Fix time calculation to handle negative nanosecond deltas

HotIce0 opened the pull request:
Fix time calculation to handle negative nanosecond deltas for device re-enumerate

In the original code, the calculation of elapsed time could produce incorrect results when the nanosecond difference (delta_nsec) was negative. This situation occurs when the current time's nanoseconds are less than the start time's nanoseconds. 

The fix introduces a check for negative delta_nsec. If delta_nsec is negative, we adjust delta_sec by decrementing it by 1 and adding NSEC_PER_SEC to delta_nsec to ensure the elapsed time is calculated correctly.

This change ensures accurate time measurement in all scenarios.

seanm commented:
Can you reproduce this occurring? How?

usbi_get_monotonic_time() is supposed to return monotonically increasing time. It should be impossible for a subsequent call to give an *earlier* time.  If somehow it does, the root cause should be investigated.  I don't think just adding a check like you've done is the right solution.

HotIce0 commented:
> Can you reproduce this occurring? How?
> 
> usbi_get_monotonic_time（） 应该返回单调递增的时间。后续调用应该不可能给出_更早_的时间。如果以某种方式出现，则应调查根本原因。我不认为像您所做的那样仅仅添加检查是正确的解决方案。

HotIce0 commented:
> Can you reproduce this occurring? How?
> 
> usbi_get_monotonic_time() is supposed to return monotonically increasing time. It should be impossible for a subsequent call to give an _earlier_ time. If somehow it does, the root cause should be investigated. I don't think just adding a check like you've done is the right solution.

I will show you the test code with added logging and the reproduced logs. You will understand once you see them. This is not related to monotonicity.

```c
static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, bool capture) {
  struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
  unsigned long claimed_interfaces = dev_handle->claimed_interfaces;
  uint8_t active_config = dpriv->active_config;
  UInt32 options = 0;
  IOUSBDeviceDescriptor descriptor;
  IOUSBConfigurationDescriptorPtr cached_configuration;
  IOUSBConfigurationDescriptor *cached_configurations;
  IOReturn kresult;
  UInt8 i;

  struct libusb_context *ctx = HANDLE_CTX (dev_handle);

  if (dpriv->in_reenumerate) {
    /* ack, two (or more) threads are trying to reset the device! abort! */
    return LIBUSB_ERROR_NOT_FOUND;
  }

  dpriv->in_reenumerate = true;

  /* store copies of descriptors so they can be compared after the reset */
  memcpy (&descriptor, &dpriv->dev_descriptor, sizeof (descriptor));
  cached_configurations = alloca (sizeof (*cached_configurations) * descriptor.bNumConfigurations);

  for (i = 0 ; i < descriptor.bNumConfigurations ; ++i) {
    (*dpriv->device)->GetConfigurationDescriptorPtr (dpriv->device, i, &cached_configuration);
    memcpy (cached_configurations + i, cached_configuration, sizeof (cached_configurations[i]));
  }

  /* if we need to release capture */
  if (get_running_version() >= 101000) {
    if (capture) {
#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
      options |= kUSBReEnumerateCaptureDeviceMask;
#endif
    }
  } else {
    capture = false;
  }

  /* from macOS 10.11 ResetDevice no longer does anything so just use USBDeviceReEnumerate */
  kresult = (*dpriv->device)->USBDeviceReEnumerate (dpriv->device, options);
  if (kresult != kIOReturnSuccess) {
    usbi_err (ctx, "USBDeviceReEnumerate: %s", darwin_error_str (kresult));
    dpriv->in_reenumerate = false;
    return darwin_to_libusb (kresult);
  }

  /* capture mode does not re-enumerate but it does require re-open */
  if (capture) {
    usbi_dbg (ctx, "darwin/reenumerate_device: restoring state...");
    dpriv->in_reenumerate = false;
    return darwin_restore_state (dev_handle, active_config, claimed_interfaces);
  }

  usbi_dbg (ctx, "darwin/reenumerate_device: waiting for re-enumeration to complete...");

  struct timespec start;
  usbi_get_monotonic_time(&start);

  while (dpriv->in_reenumerate) {
    struct timespec delay = {.tv_sec = 0, .tv_nsec = 1000};
    nanosleep (&delay, NULL);

    struct timespec now;
    usbi_get_monotonic_time(&now);
    long delta_sec = now.tv_sec - start.tv_sec;
    long delta_nsec = now.tv_nsec - start.tv_nsec;
    if (delta_nsec < 0) {
      usbi_err(ctx, "zg: s(%ld %ld) e(%ld %ld)", start.tv_sec, start.tv_nsec,
               now.tv_sec, now.tv_nsec);
      usbi_err(ctx, "zg: delta_sec %ld delta_nsec %ld", delta_sec, delta_nsec);
      delta_sec -= 1;
      delta_nsec += NSEC_PER_SEC;
      usbi_err(ctx, "zg: fixed delta_sec %ld delta_nsec %ld", delta_sec, delta_nsec);
    }
    
    unsigned long long elapsed_us = (unsigned long long)delta_sec * USEC_PER_SEC +
                                    (unsigned long long)delta_nsec / 1000ULL;

    if (elapsed_us >= DARWIN_REENUMERATE_TIMEOUT_US) {
      usbi_err (ctx, "darwin/reenumerate_device: timeout waiting for reenumerate");
      dpriv->in_reenumerate = false;
      return LIBUSB_ERROR_TIMEOUT;
    }
  }

  /* compare descriptors */
  usbi_dbg (ctx, "darwin/reenumerate_device: checking whether descriptors changed");

  if (memcmp (&descriptor, &dpriv->dev_descriptor, sizeof (descriptor)) != 0) {
    /* device descriptor changed. need to return not found. */
    usbi_dbg (ctx, "darwin/reenumerate_device: device descriptor changed");
    return LIBUSB_ERROR_NOT_FOUND;
  }

  for (i = 0 ; i < descriptor.bNumConfigurations ; ++i) {
    (void) (*dpriv->device)->GetConfigurationDescriptorPtr (dpriv->device, i, &cached_configuration);
    if (memcmp (cached_configuration, cached_configurations + i, sizeof (cached_configurations[i])) != 0) {
      usbi_dbg (ctx, "darwin/reenumerate_device: configuration descriptor %d changed", i);
      return LIBUSB_ERROR_NOT_FOUND;
    }
  }

  usbi_dbg (ctx, "darwin/reenumerate_device: device reset complete. restoring state...");

  return darwin_restore_state (dev_handle, active_config, claimed_interfaces);
}
```

[libusb_debug_log.txt](https://github.com/user-attachments/files/18944472/libusb_debug_log.txt)

HotIce0 commented:
[ 0.360327] [00571c28] libusb: error [darwin_reenumerate_device] zg: s(370548 950204000) e(370549 207573000)
[ 0.360354] [00571c28] libusb: error [darwin_reenumerate_device] zg: delta_sec 1 delta_nsec -742631000
[ 0.360400] [00571c28] libusb: error [darwin_reenumerate_device] zg: fixed delta_sec 0 delta_nsec 257369000

HotIce0 commented:
tv_nsec is not monotonically increasing; it is periodic, so the calculated delta_nsec can be negative.

HotIce0 commented:
I found the TIMESPEC_SUB macro in libusbi.h. I'm going to use this macro for implementation instead. I'll verify it after 10 hours and then give feedback on the results at that time. @seanm

seanm commented:
Ah, I see what you mean now!  Nice catch.

Yes, using TIMESPEC_SUB is a good solution.

We should probably also search the codebase for `tv_nsec` and look for similar mistakes...

HotIce0 commented:
> Ah, I see what you mean now! Nice catch.
> 
> Yes, using TIMESPEC_SUB is a good solution.
> 
> We should probably also search the codebase for `tv_nsec` and look for similar mistakes...

I did a global search and check, and it seems that the same problem also exists in measure of examples/sam3u_benchmark.c.

Do we need to open a new pull request for the merge?

https://github.com/libusb/libusb/blob/de38189e8014fa393f4d8c1d9d3fdf5e2a95899d/examples/sam3u_benchmark.c#L164

original
```c
static void measure(void)
{
	struct timeval tv_stop;
	unsigned long diff_msec;

	get_timestamp(&tv_stop);

	diff_msec = (tv_stop.tv_sec - tv_start.tv_sec) * 1000L;
	diff_msec += (tv_stop.tv_usec - tv_start.tv_usec) / 1000L;

	printf("%lu transfers (total %lu bytes) in %lu milliseconds => %lu bytes/sec\n",
		num_xfer, num_bytes, diff_msec, (num_bytes * 1000L) / diff_msec);
}
```

fixed
```c
static void measure(void)
{
	struct timeval tv_stop;
	unsigned long diff_msec;

	get_timestamp(&tv_stop);

	long diff_sec = tv_stop.tv_sec - tv_start.tv_sec;
	long diff_usec = tv_stop.tv_usec - tv_start.tv_usec;
	if (diff_usec < 0) {
		diff_sec -= 1;
		diff_usec += NSEC_PER_SEC;
	}
	diff_msec = (unsigned long)diff_sec * 1000UL;
	diff_msec += (unsigned long)diff_usec / 1000UL;

	printf("%lu transfers (total %lu bytes) in %lu milliseconds => %lu bytes/sec\n",
		num_xfer, num_bytes, diff_msec, (num_bytes * 1000L) / diff_msec);
}
```


test
```c
#include <cstdio>

struct timeval
{
	long tv_sec;
	long tv_usec;
};

void original(long sec1, long usec1, long sec2, long usec2)
{
	unsigned long diff_msec;
	struct timeval tv_start, tv_stop;
	tv_start.tv_sec = sec2;
	tv_start.tv_usec = usec2;
	tv_stop.tv_sec = sec1;
	tv_stop.tv_usec = usec1;

	diff_msec = (tv_stop.tv_sec - tv_start.tv_sec) * 1000L;
	diff_msec += (tv_stop.tv_usec - tv_start.tv_usec) / 1000L;

	printf("%s %lu\n", __FUNCTION__, diff_msec);
}

void fixed(long sec1, long usec1, long sec2, long usec2)
{
	unsigned long diff_msec;
	struct timeval tv_start, tv_stop;
	tv_start.tv_sec = sec2;
	tv_start.tv_usec = usec2;
	tv_stop.tv_sec = sec1;
	tv_stop.tv_usec = usec1;
	
	long diff_sec = tv_stop.tv_sec - tv_start.tv_sec;
	long diff_usec = tv_stop.tv_usec - tv_start.tv_usec;
	if (diff_usec < 0) {
		diff_sec -= 1;
		diff_usec += 1000000000;
	}
	diff_msec = diff_sec * 1000UL;
	diff_msec += diff_usec / 1000UL;

	printf("%s %lu\n", __FUNCTION__, diff_msec);
}

int main(void)
{
	// positive tests
	original(1 /*end sec*/, 2000 /*end usec*/, 0 /*start sec*/, 0 /*start usec*/);
	fixed(1, 2000, 0, 0);
	// nagtive tests
	original(1, 0, 0, 2000);
	fixed(1, 0, 0, 2000);
	return 0;
}

// output of test code as bellow
original 1002
fixed 1002
original 998
fixed 999998

```

seanm commented:
> Do we need to open a new pull request for the merge?

You can add to this PR.

HotIce0 commented:
> > Ah, I see what you mean now! Nice catch.
> > Yes, using TIMESPEC_SUB is a good solution.
> > We should probably also search the codebase for `tv_nsec` and look for similar mistakes...
> 
> I did a global search and check, and it seems that the same problem also exists in measure of examples/sam3u_benchmark.c.
> 
> Do we need to open a new pull request for the merge?
> 
> https://github.com/libusb/libusb/blob/de38189e8014fa393f4d8c1d9d3fdf5e2a95899d/examples/sam3u_benchmark.c#L164
> 
> original
> 
> ```c
> static void measure(void)
> {
> 	struct timeval tv_stop;
> 	unsigned long diff_msec;
> 
> 	get_timestamp(&tv_stop);
> 
> 	diff_msec = (tv_stop.tv_sec - tv_start.tv_sec) * 1000L;
> 	diff_msec += (tv_stop.tv_usec - tv_start.tv_usec) / 1000L;
> 
> 	printf("%lu transfers (total %lu bytes) in %lu milliseconds => %lu bytes/sec\n",
> 		num_xfer, num_bytes, diff_msec, (num_bytes * 1000L) / diff_msec);
> }
> ```
> 
> fixed
> 
> ```c
> #if !defined(NSEC_PER_SEC)
> #define NSEC_PER_SEC	1000000000L
> #endif
> 
> static void measure(void)
> {
> 	struct timeval tv_stop;
> 	unsigned long diff_msec;
> 
> 	get_timestamp(&tv_stop);
> 
> 	long diff_sec = tv_stop.tv_sec - tv_start.tv_sec;
> 	long diff_usec = tv_stop.tv_usec - tv_start.tv_usec;
> 	if (diff_usec < 0) {
> 		diff_sec -= 1;
> 		diff_usec += NSEC_PER_SEC;
> 	}
> 	diff_msec = (unsigned long)diff_sec * 1000UL;
> 	diff_msec += (unsigned long)diff_usec / 1000UL;
> 
> 	printf("%lu transfers (total %lu bytes) in %lu milliseconds => %lu bytes/sec\n",
> 		num_xfer, num_bytes, diff_msec, (num_bytes * 1000L) / diff_msec);
> }
> ```
> 
> test
> 
> ```c
> #include <cstdio>
> 
> struct timeval
> {
> 	long tv_sec;
> 	long tv_usec;
> };
> 
> void original(long sec1, long usec1, long sec2, long usec2)
> {
> 	unsigned long diff_msec;
> 	struct timeval tv_start, tv_stop;
> 	tv_start.tv_sec = sec2;
> 	tv_start.tv_usec = usec2;
> 	tv_stop.tv_sec = sec1;
> 	tv_stop.tv_usec = usec1;
> 
> 	diff_msec = (tv_stop.tv_sec - tv_start.tv_sec) * 1000L;
> 	diff_msec += (tv_stop.tv_usec - tv_start.tv_usec) / 1000L;
> 
> 	printf("%s %lu\n", __FUNCTION__, diff_msec);
> }
> 
> void fixed(long sec1, long usec1, long sec2, long usec2)
> {
> 	unsigned long diff_msec;
> 	struct timeval tv_start, tv_stop;
> 	tv_start.tv_sec = sec2;
> 	tv_start.tv_usec = usec2;
> 	tv_stop.tv_sec = sec1;
> 	tv_stop.tv_usec = usec1;
> 	
> 	long diff_sec = tv_stop.tv_sec - tv_start.tv_sec;
> 	long diff_usec = tv_stop.tv_usec - tv_start.tv_usec;
> 	if (diff_usec < 0) {
> 		diff_sec -= 1;
> 		diff_usec += 1000000000;
> 	}
> 	diff_msec = diff_sec * 1000UL;
> 	diff_msec += diff_usec / 1000UL;
> 
> 	printf("%s %lu\n", __FUNCTION__, diff_msec);
> }
> 
> int main(void)
> {
> 	// positive tests
> 	original(1 /*end sec*/, 2000 /*end usec*/, 0 /*start sec*/, 0 /*start usec*/);
> 	fixed(1, 2000, 0, 0);
> 	// nagtive tests
> 	original(1, 0, 0, 2000);
> 	fixed(1, 0, 0, 2000);
> 	return 0;
> }
> 
> // output of test code as bellow
> original 1002
> fixed 1002
> original 998
> fixed 999998
> ```

@seanm The Darwin code has been successfully tested on my macOS terminal. As for the sam3u_benchmark code, since I do not have the device (16c0:0763), I have only verified its correctness using the previous test code. What else do I need to do for this pull request?

HotIce0 commented:
@mcuee I need some help with reviewing the this PR. Thand you.
This PR is not related to the platform.

mcuee commented:
> @mcuee I need some help with reviewing the this PR. Thand you. This PR is not related to the platform.

Unfortunately I am not able to review this properly myself (more in the role of testing and supporting as I do not know much about programming myself. Basically I cannot code and I can only review simple codes). Need the review from either @tormodvolden or @hjelmn.

tormodvolden commented:
> I can take a look. I am trying to find more time to actively work on libusb. Should have time tomorrow to look at all open PRs.

That's great @hjelmn ! I don't know if you already saw it, we are planning a release with what is in master now, before merging anything new. I'll roll a RC maybe today.

mcuee commented:
@hjelmn 

Ping. Thanks.

HotIce0 commented:
@hjelmn

tormodvolden commented:
Fixes #1637. A part from the version_nano.h change, this looks good.

tormodvolden commented:
@HotIce0 Is your full name correctly written "Zeng Guang"?

HotIce0 commented:
> @HotIce0 Is your full name correctly written "Zeng Guang"?
yes

HotIce0 committed:
darwin: Fix time calculation for negative nanosecond deltas to resolve re-enumerate timeout

HotIce0 committed:
sam3u_benchmark: Fix time calculation for negative nanosecond deltas

HotIce0 commented:
> Fixes #1637. A part from the version_nano.h change, this looks good.

I directly deleted the updated part of the version_nano.h file—it's too prone to conflicts since it keeps changing all the time.

tormodvolden commented:
Thanks, yes that file shouldn't be changed in pull requests.

sonatique reviewed (commented):
Nitpicking: technically the commit "[sam3u_benchmark: Fix time calculation for negative nanosecond deltas](https://github.com/libusb/libusb/pull/1610/commits/713b6e5274482543fd36068a78d3fffe65b22c60)" fixes microsecond calxculations not nanoseconds (as opposed to the other one) contrary to what the comment says.

tormodvolden commented:
I have merged the Darwin regression fix. @HotIce0 Thanks for your contribution!

However, I think the sam3u_benchmark change doesn't fix anything. The original code was correct. It looks like microseconds and nanoseconds got mixed up in the test case. (The test case is a great idea though!) If you fix up the `diff_usec += 1000000000;` typo the test case will succeed. Your change only affects sub-millisecond deltas, and changes from rounding down to rounding up.

seanm commented:
Ideally it should just use `TIMESPEC_SUB` and not reinvent the wheel.

mcuee commented:
From here: https://github.com/libusb/libusb/issues/1635#issuecomment-2907868626

> That's great. However, that commit only prevents timeout. It is possible that the absence of timeout now avoids the issue causing the segfault in this report. But if libusb still may segfault when there is a timeout or other device timing issue, there must be a vulnerability that we should try to fix IMO. libusb should never segfault, whatever timing and device behavior is taking place.

tormodvolden commented:
> I can take a look. I am trying to find more time to actively work on libusb. Should have time tomorrow to look at all open PRs.

@hjelmn If you find some spare cycles, I would particularly appreciate if you give your opinion on suggested API changes (https://github.com/libusb/libusb/labels/API) where you haven't already. You don't necessarily need to look at the implementations, just the API itself, if it makes sense and seems the best way forward. Most times we have people reviewing the code but the really hard question is how do we want our API to evolve. You have been developing and shaping the libusb API for many years and I hope you will continue to be involved for long.

seanm commented:
This was fixed in 1.0.29.

mcuee commented:
@HotIce0 

I will close this PR for now. If you want to improve the second commit, please start a new PR.
